fvm_actor_utils/src/receiver/mod.rs

```diff
@@ -0,0 +1,229 @@
+}
+
+/// Type of asset received - could be tokens (FRC46 or other) or other assets
+pub type ReceiverType = u32;
+
+#[derive(Error, Debug)]
+pub enum ReceiverHookError {
+    #[error("receiver hook was not called")]
+    NotCalled,
+    #[error("receiver hook was already called")]
+    AlreadyCalled,
+    #[error("error encoding to ipld")]
+    IpldEncoding(#[from] fvm_ipld_encoding::Error),
+    #[error("error sending message")]
+    Messaging(#[from] MessagingError),
+    #[error("receiver hook error from {address:?} when called with {receiver_params:?}: exit_code={exit_code:?}, return_data={return_data:?}")]
+    Receiver {
+        address: Address,
+        exit_code: ExitCode,
+        return_data: RawBytes,
+        receiver_params: UniversalReceiverParams,
+    },
+}
+
+impl From<&ReceiverHookError> for ExitCode {
+    fn from(error: &ReceiverHookError) -> Self {
+        match error {
+            ReceiverHookError::NotCalled | ReceiverHookError::AlreadyCalled => {
+                ExitCode::USR_ASSERTION_FAILED
+            }
+            ReceiverHookError::IpldEncoding(_) => ExitCode::USR_SERIALIZATION,
+            ReceiverHookError::Receiver {
+                address: _,
+                return_data: _,
+                receiver_params: _,
+                exit_code,
+            } => *exit_code,
+            ReceiverHookError::Messaging(e) => e.into(),
+        }
+    }
+}
+
+/// Parameters for universal receiver
+///
+/// Actual payload varies with asset type
+/// eg: FRC46_TOKEN_TYPE will come with a payload of FRC46TokenReceived
+#[derive(Serialize_tuple, Deserialize_tuple, PartialEq, Eq, Clone, Debug)]
+pub struct UniversalReceiverParams {
+    /// Asset type
+    pub type_: ReceiverType,
+    /// Payload corresponding to asset type
+    pub payload: RawBytes,
+}
+impl Cbor for UniversalReceiverParams {}
+
+pub trait RecipientData {
+    fn set_recipient_data(&mut self, data: RawBytes);
+}
+
+/// Implements a guarded call to a token receiver hook
+///
+/// Mint and Transfer operations will return this so that state can be updated and saved
+/// before making the call into the receiver hook.
+///
+/// This also tracks whether the call has been made or not, and
+/// will panic if dropped without calling the hook.
+#[derive(Debug)]
+pub struct ReceiverHook<T: RecipientData> {
+    address: Address,
+    token_type: ReceiverType,
+    token_params: RawBytes,
+    called: bool,
+    result_data: Option<T>,
+}
+
+impl<T: RecipientData> ReceiverHook<T> {
+    /// Construct a new ReceiverHook call
+    pub fn new(
+        address: Address,
+        token_params: RawBytes,
+        token_type: ReceiverType,
+        result_data: T,
+    ) -> Self {
+        ReceiverHook {
+            address,
+            token_params,
+            token_type,
+            called: false,
+            result_data: Some(result_data),
+        }
+    }
+
+    /// Construct a new FRC46 ReceiverHook call
+    pub fn new_frc46(
```

Comment:
If we find that we need more of these helper/shortcut functions, we could implement them in traits instead. So we'd just `use frc46_token::ReceiverHook` (for example) and get frc46-specific helper functions. I don't think it's something we need to do for now but worth keeping in mind as hook usage expands into NFTs and beyond.